### ISSUE TITLE
How does neat-python ensure the uniqueness of historical markers? 

### ISSUE THREAD

**Issue by sopotc** [CodeReclaimers/neat-python#285]
Looking at the connection genes, they do not keep track of any global historical marker.

Apparently the only thing they're keyed on are the start nodes and end nodes. A (start node, end node) key is by no means a guarantee that holds the property of global innovation markers, as they can mean different structures in different genomes. For example node 4 in genome 1 can be in between node 2 and 3, but in genome 2 it can be between 1 and 5. So relying on node ids for speciation and crossover compromises the key property of the NEAT algorithm - the fact that the same historical marker means the same structural mutation _across the population_. 

In other words, if connection gene A is considered to be the same as connection gene B, because their (start, end) tuple is the same, then this is effectively a case of competing conventions which then means the algorithm implemented anything but NEAT. The whole reason NEAT added the historical markers is to avoid exactly this problem.

**Comment by sopotc** [comment]
See section 3.2 in the original [paper](https://nn.cs.utexas.edu/downloads/papers/stanley.ec02.pdf). The whole point of the historical marker is to switch the 'coordinates' of a gene from the (start, end) formulation to a global (innovation) coordinate. Thus two genes are the same only when they have the same innovation, and not when they have the same (start, end). Consequently, during crossover they're lined up by their innovation number, and this innovation number is also used during speciation distance calculations.

Why introduce innovation numbers (see fig. 2 and fig. 3 in original paper) when you already have the (in, out) keys?

**Comment by th555** [comment]
There used to be a `__global_innov_number` in the connection gene, the most recent commit that still has it is [here](https://github.com/CodeReclaimers/neat-python/blob/803902d81464b6dcbaff85d3ea9dd9bae55cbf78/neat/genome.py#L101). Then this got moved to a separate, but still global, indexer object ([here](https://github.com/CodeReclaimers/neat-python/blob/dc51301d046c4828aac0c80b86bef47bf5d79e3e/neat/genes.py#L102)). Looks like global innovation numbers fell by the wayside in [this large refactor](https://github.com/CodeReclaimers/neat-python/commit/8de3e6bc7a7e5b6962e685264f07001a09f84592).

**Comment by CodeReclaimers** [comment]
@sopotc Thanks for pointing this out! I'm restoring the innovation number tracking for version 1.0.
